cmd/mailchain/internal/settings/mailbox_state.go

```diff
@@ -1,44 +1,60 @@
 package settings
 
 import (
+	"fmt"
+
 	"github.com/mailchain/mailchain/cmd/mailchain/internal/settings/defaults"
 	"github.com/mailchain/mailchain/cmd/mailchain/internal/settings/output"
 	"github.com/mailchain/mailchain/cmd/mailchain/internal/settings/values"
 	"github.com/mailchain/mailchain/stores"
+	"github.com/mailchain/mailchain/stores/bdbstore"
 	"github.com/mailchain/mailchain/stores/ldbstore"
-	"github.com/pkg/errors"
 )
 
+// State interface wraps basic methods required for mailbox state.
+type State interface {
+	Produce() (stores.State, error)
+	Output() output.Element
+	Kind() values.String
+}
+
+// mailboxState returns the default MailboxState which uses leveldb store.
 func mailboxState(s values.Store) *MailboxState {
-	k := &MailboxState{
-		Kind:                values.NewDefaultString(defaults.MailboxStateKind, s, "mailboxState.kind"),
-		mailboxStateLevelDB: mailboxStateLevelDB(s),
+	return &MailboxState{
+		mailboxState: mailboxStateLevelDB(s),
+	}
+}
+
+// mailboxStateFromKind returns MailboxState based on store kind.
+func mailboxStateFromKind(kind string, s values.Store) (*MailboxState, error) {
+	switch kind {
+	case StoreLevelDB:
+		m := &MailboxState{mailboxState: mailboxStateLevelDB(s)}
+		return m, nil
+	case StoreBadgerDB:
+		m := &MailboxState{mailboxState: mailboxStateBadgerDB(s)}
+		return m, nil
+	default:
+		return nil, fmt.Errorf("unknown state store %q", kind)
 	}
-	return k
 }
 
 // MailboxState settings for mailbox state storage
 type MailboxState struct {
-	Kind                values.String
-	mailboxStateLevelDB MailboxStateLevelDB
+	mailboxState State
 }
 
 // Produce `stores.State` based on configuration settings.
 func (s MailboxState) Produce() (stores.State, error) {
-	switch s.Kind.Get() {
-	case StoreLevelDB:
-		return s.mailboxStateLevelDB.Produce()
-	default:
-		return nil, errors.Errorf("%q is an unsupported mailbox state", s.Kind.Get())
-	}
```

Comment:
This should be closer to the older version of this function and add as case statement for `StoreBadgerDB` This way it will inspect `Kind` to decide what datastore to return